From 0b15aee9d446a0cbf1369783ec76109573c624bf Mon Sep 17 00:00:00 2001
From: Zhenggang Luo <zhenggang.luo@amlogic.com>
Date: Thu, 29 Jan 2015 23:51:11 +0800
Subject: [PATCH 5384/5965] PD #101059: change i2s/iec958 couple method

1-when non pcm_output,let iec958/i2s use same pcm_buf_addr;
2-when rawoutput,let alsa control them individually;
3-reinit iec958 when detect AIU_MEM_IEC958_START_PTR!=
 AIU_MEM_I2S_START_PTR when pcm_output;

Change-Id: I6b80eaa1c10598af051ce590aa3be0991ae5cd4e
---
 drivers/amlogic/amaudio/amaudio.c          |  5 +++
 drivers/amlogic/audiodsp/audiodsp_module.c | 14 +++++++--
 sound/soc/aml/m8/aml_i2s.c                 | 36 ++++++++++++++++++++++
 sound/soc/aml/m8/aml_i2s_dai.c             | 10 ++++++
 4 files changed, 63 insertions(+), 2 deletions(-)

diff --git a/drivers/amlogic/amaudio/amaudio.c b/drivers/amlogic/amaudio/amaudio.c
index 7a814380d614..6a640ed2405a 100755
--- a/drivers/amlogic/amaudio/amaudio.c
+++ b/drivers/amlogic/amaudio/amaudio.c
@@ -107,6 +107,9 @@ int resample_delta=0;
 EXPORT_SYMBOL(resample_delta);
 extern unsigned int timestamp_enable_resample_flag;
 extern unsigned int timestamp_resample_type_flag;
+int kernel_android_50=0;
+EXPORT_SYMBOL(kernel_android_50);
+
 //--------------------------------------------
 #define DEBUG_DUMP 1
 
@@ -2034,6 +2037,8 @@ static ssize_t store_debug(struct class* class, struct class_attribute* attr,  c
            memset(dtsm6_apres_assets_Array,0,sizeof(dtsm6_apres_assets_Array));
      }else if(strncmp(buf, "dtsm6_hps_hint",14)==0){
            dtsm6_HPS_hint=simple_strtoul(buf+14,&after,10);;
+     }else if(strncmp(buf, "kernel_android_50",17)==0){
+           kernel_android_50=1;
      }
      return count;
 }
diff --git a/drivers/amlogic/audiodsp/audiodsp_module.c b/drivers/amlogic/audiodsp/audiodsp_module.c
index 6d1c34c54a74..c4d00166831a 100755
--- a/drivers/amlogic/audiodsp/audiodsp_module.c
+++ b/drivers/amlogic/audiodsp/audiodsp_module.c
@@ -804,10 +804,14 @@ static ssize_t digital_codec_show(struct class*cla, struct class_attribute* attr
   pbuf += sprintf(pbuf, "%d\n",IEC958_mode_codec);
   return (pbuf-buf);
 }
+
+extern void aml_audio_hw_trigger(void);
+extern int kernel_android_50;
 static ssize_t digital_codec_store(struct class* class, struct class_attribute* attr,
    const char* buf, size_t count )
 {
 	int digital_codec  = 0;
+	unsigned mode_codec = IEC958_mode_codec;
 	if(buf){
   		digital_codec = simple_strtoul(buf, NULL, 10);
 		if(digital_codec < SUPPORT_TYPE_NUM){
@@ -817,8 +821,14 @@ static ssize_t digital_codec_store(struct class* class, struct class_attribute*
 		else{
 			printk("IEC958 type set exceed supported range \n");
 		}
-	}		
-  return count;
+	}
+	// when raw output switch to pcm output ,need trigger pcm hw prepare to re-init hw configuration
+	printk("last mode %d,now %d \n",mode_codec,IEC958_mode_codec);
+	if (mode_codec && !IEC958_mode_codec && kernel_android_50==1) {
+	    printk("[kernel_android_50/%d]trigger underrun\n",kernel_android_50);
+	    aml_audio_hw_trigger();
+	}
+	return count;
 }
 static ssize_t print_flag_show(struct class*cla, struct class_attribute* attr, char* buf)
 {
diff --git a/sound/soc/aml/m8/aml_i2s.c b/sound/soc/aml/m8/aml_i2s.c
index 1678e51e0ad0..56105ca9910e 100755
--- a/sound/soc/aml/m8/aml_i2s.c
+++ b/sound/soc/aml/m8/aml_i2s.c
@@ -68,6 +68,13 @@ EXPORT_SYMBOL(aml_i2s_playback_phy_start_addr);
 EXPORT_SYMBOL(aml_i2s_capture_phy_start_addr);
 EXPORT_SYMBOL(aml_i2s_alsa_write_addr);
 
+static int trigger_underrun = 0;
+void aml_audio_hw_trigger(void)
+{
+    trigger_underrun = 1;
+}
+EXPORT_SYMBOL(aml_audio_hw_trigger);
+
 
 /*--------------------------------------------------------------------------*\
  * Hardware definition
@@ -264,7 +271,16 @@ static int aml_i2s_hw_free(struct snd_pcm_substream *substream)
 
 static int aml_i2s_prepare(struct snd_pcm_substream *substream)
 {
+
+	struct snd_pcm_runtime *runtime = substream->runtime;
+	struct aml_runtime_data *prtd = runtime->private_data;
+	audio_stream_t *s = &prtd->s;
+
 	ALSA_TRACE();
+	if(s && s->device_type == AML_AUDIO_I2SOUT && trigger_underrun){
+	     printk("clear i2s out trigger underrun \n");
+	     trigger_underrun = 0;
+	}
 	return 0;
 }
 
@@ -581,6 +597,11 @@ static int aml_i2s_close(struct snd_pcm_substream *substream)
 	return 0;
 }
 
+extern unsigned int IEC958_mode_codec;
+extern void audio_out_i2s_enable(unsigned flag);
+extern void aml_hw_iec958_init(struct snd_pcm_substream *substream);
+extern void audio_hw_958_enable(unsigned flag);
+extern int kernel_android_50;
 
 static int aml_i2s_copy_playback(struct snd_pcm_runtime *runtime, int channel,
 		    snd_pcm_uframes_t pos,
@@ -596,6 +617,21 @@ static int aml_i2s_copy_playback(struct snd_pcm_runtime *runtime, int channel,
     struct aml_audio_buffer *tmp_buf = buffer->private_data;
     void *ubuf = tmp_buf->buffer_start;
     audio_stream_t *s = &prtd->s;
+    int force_reinit_958=0;
+	force_reinit_958= (IEC958_mode_codec==0 && (READ_MPEG_REG(AIU_MEM_IEC958_START_PTR)!=READ_MPEG_REG(AIU_MEM_I2S_START_PTR)));
+	if (s && s->device_type == AML_AUDIO_I2SOUT && (trigger_underrun) && kernel_android_50==1 ){
+	    printk("i2s out trigger underrun force_reinit_958/%d trigger_underrun/%d IEC958_mode_codec/%d IEC958_START_PTR/0x%x I2S_START_PTR/0x%x\n",
+			force_reinit_958,trigger_underrun,IEC958_mode_codec,READ_MPEG_REG(AIU_MEM_IEC958_START_PTR),READ_MPEG_REG(AIU_MEM_I2S_START_PTR));
+	    //trigger_underrun = 0;
+	    return -EFAULT;
+	}
+	if (s && s->device_type == AML_AUDIO_I2SOUT && force_reinit_958 && kernel_android_50==1){
+		printk("i2s out trigger underrun force_reinit_958/%d",force_reinit_958);
+		audio_hw_958_enable(0);
+		aml_hw_iec958_init(substream);
+		audio_hw_958_enable(1);
+	}
+
     if(s->device_type == AML_AUDIO_I2SOUT){
         aml_i2s_alsa_write_addr = frames_to_bytes(runtime, pos);
     }
diff --git a/sound/soc/aml/m8/aml_i2s_dai.c b/sound/soc/aml/m8/aml_i2s_dai.c
index 614da2792c38..e7c317653c7f 100755
--- a/sound/soc/aml/m8/aml_i2s_dai.c
+++ b/sound/soc/aml/m8/aml_i2s_dai.c
@@ -38,6 +38,10 @@ static int i2s_pos_sync = 0;
 #define ALSA_TRACE()   
 #endif
 extern int amaudio2_enable;
+extern unsigned int IEC958_mode_codec;
+extern int kernel_android_50;
+
+
 extern int set_i2s_iec958_samesource(int enable);
 
 static int i2sbuf[32+16];
@@ -195,6 +199,8 @@ static int aml_dai_i2s_prepare(struct snd_pcm_substream *substream,
         if(amaudio2_enable == 1){
             aml_hw_iec958_init(substream);
             set_i2s_iec958_samesource(1);
+        }else if(IEC958_mode_codec==0 && kernel_android_50==1){
+            aml_hw_iec958_init(substream);
         }
     }
     if(runtime->channels == 8){
@@ -220,6 +226,8 @@ static int aml_dai_i2s_trigger(struct snd_pcm_substream *substream, int cmd,
 				if(amaudio2_enable == 1){
 					audio_hw_958_enable(1);
 					set_i2s_iec958_samesource(1);
+				}else if(IEC958_mode_codec==0 && kernel_android_50==1){
+					audio_hw_958_enable(1);
 				}
 			}else{
 				audio_in_i2s_enable(1);
@@ -237,6 +245,8 @@ static int aml_dai_i2s_trigger(struct snd_pcm_substream *substream, int cmd,
 				if(amaudio2_enable == 1){
 					audio_hw_958_enable(0);
 					set_i2s_iec958_samesource(0);
+				}else if(IEC958_mode_codec==0 && kernel_android_50==1){
+					audio_hw_958_enable(0);
 				}
 			}else{
 				audio_in_i2s_enable(0);
-- 
2.19.0

